lib/core/checker/checker.cc

```diff
@@ -10,6 +19,263 @@
+      return !(*this)(message.second);
+    });
+  }
+
+  bool operator()(const ast::Message &ast_message) {
+    AddScope();
+    if (!IsInScope(ast_message.name)) {
+      return false;
+    }
+    if (!(*this)(ast_message.type_dependencies)) {
+      return false;
+    }
+    if (!(*this)(ast_message.fields)) {
+      return false;
+    }
+    DeleteScope();
+
+    return true;
+  }
+
+  bool operator()(const std::vector<ast::TypedVariable> &dependencies) {
+    return !std::ranges::any_of(dependencies.begin(), dependencies.end(), [this](auto &dependency) {
+      return !(*this)(dependency);
+    });
+  }
+
+  bool operator()(const ast::TypedVariable &typed_variable) {
+    if (!(*this)(typed_variable.type_expression)) {
+      return false;
+    }
+    return AddName(typed_variable.name);
+  }
+
+  bool operator()(const ast::TypeExpression &type_expression) {
+    if (!IsInScope(type_expression.name)) {
+      return false;
+    }
+
+    return (*this)(type_expression.parameters);
+  }
+
+  bool operator()(const std::vector<std::unique_ptr<ast::Expression>> &expressions) {
+    for (auto &expression : expressions) {
+      if (std::visit(*this, *expression)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool operator()(const ast::ConstructedValue &value) {
+    if (!IsInScope(value.constructor_identifier)) {
+      return false;
+    }
+    AddScope();
+    if (!(*this)(value.constructor_identifier, value.fields)) {
+      return false;
+    }
+    DeleteScope();
+
+    return true;
+  }
+
+  bool operator()(
+      InternedString constructor_identifier,
+      const std::vector<std::pair<InternedString, std::unique_ptr<ast::Expression>>> &fields) {
+    for (auto &field : fields) {
+      if (!constructors_fields_[constructor_identifier].contains(field.first)) {
+        return false;
+      }
+
+      if (!(*this)(field)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool operator()(const std::pair<InternedString, std::unique_ptr<ast::Expression>> &field) {
+    if (!std::visit(*this, *field.second)) {
+      return false;
+    }
+    AddName(field.first);
+  }
+  template <typename T>
+  bool operator()(const ast::ScalarValue<T> &value) {
+    return true;
+  }
+
+  bool operator()(const ast::BinaryExpression &expr) const {
+    return std::visit(*this, *(expr.left)) && std::visit(*this, *expr.right);
+  }
+  bool operator()(const ast::UnaryExpression &expr) const {
+    return std::visit(*this, *(expr.expression));
+  }
+
+  bool operator()(const ast::VarAccess &var_access) {
+    return IsInScope(var_access.var_identifier);
+  }
+
+private:
+  bool isShadowing_ = false;
+
+  std::deque<std::unordered_set<InternedString>> scopes_;
+  std::unordered_map<InternedString, std::unordered_set<InternedString>> constructors_fields_;
+  std::unordered_map<InternedString, std::unordered_set<InternedString>> messages_fields_;
```

Comment:
I think we can get rid of this map (because we don't use it anyway) but add the fields of messages as if they had the eponymous constructor